src/cryptonote_core/service_node_voting.h

```diff
@@ -125,27 +125,38 @@ namespace service_nodes
     // TODO(loki): Review relay behaviour and all the cases when it should be triggered
     void                         set_relayed         (const std::vector<quorum_vote_t>& votes);
     void                         remove_expired_votes(uint64_t height);
-    void                         remove_used_votes   (std::vector<cryptonote::transaction> const &txs);
+    void                         remove_used_votes   (std::vector<cryptonote::transaction> const &txs, uint8_t hard_fork_version);
     std::vector<quorum_vote_t>   get_relayable_votes () const;
 
-  private:
-    struct deregister_pool_entry
+    struct state_change_pool_key
     {
-      deregister_pool_entry(uint64_t height, uint32_t worker_index): height(height), worker_index(worker_index) {}
+      state_change_pool_key(const quorum_vote_t &vote)
+          : height{vote.block_height}, worker_index{vote.state_change.worker_index}, state{vote.state_change.state} {}
+      state_change_pool_key(const cryptonote::tx_extra_service_node_state_change &sc)
+          : height{sc.block_height}, worker_index{sc.service_node_index}, state{sc.state} {}
+
       uint64_t                     height;
       uint32_t                     worker_index;
-      std::vector<pool_vote_entry> votes;
+      new_state                    state;
+
+      bool operator==(const state_change_pool_key &e) const { return height == e.height && worker_index == e.worker_index && state == e.state; }
+      struct hash { size_t operator()(const state_change_pool_key &e) const { return std::hash<uint64_t>{}(e.height) + e.worker_index; } };
     };
-    std::vector<deregister_pool_entry> m_deregister_pool;
 
-    struct checkpoint_pool_entry
+  private:
+    std::unordered_map<state_change_pool_key, std::vector<pool_vote_entry>, state_change_pool_key::hash> m_state_change_pool;
```

Comment:
An unordered map on a crypto hash is actually extremely efficient in our particular case: it doesn't have the usual overhead of having to calculate a hash because the first values of the key itself is already a perfectly valid hash.

Reply:
Actually, scratch that, I'm wrong about it in this case (it does hash on a few fields).  I'll revert it to a vector.

Reply:
Switched back to vector.